cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/replication/ReplicationListener.java

```diff
@@ -0,0 +1,75 @@
+package com.cloudant.sync.internal.replication;
+
+import com.cloudant.sync.event.Subscribe;
+import com.cloudant.sync.event.notifications.ReplicationCompleted;
+import com.cloudant.sync.event.notifications.ReplicationErrored;
+import com.cloudant.sync.replication.ReplicationPolicyManager;
+import com.cloudant.sync.replication.Replicator;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This class is not intended as API, it is public for EventBus access only.
+ * API consumers should not call these methods. See #setReplicationsCompletedListener for more
+ * information on replication lifecycle events.
+ *
+ * @api_private
+ */
+public class ReplicationListener {
+
+    private final Set<Replicator> replicatorsInProgress = new HashSet<Replicator>();
+    private ReplicationPolicyManager.ReplicationsCompletedListener mReplicationsCompletedListener
+            = null;
+
+    public void setReplicationsCompletedListener(ReplicationPolicyManager
+                                                         .ReplicationsCompletedListener
+                                                         mReplicationsCompletedListener) {
+        this.mReplicationsCompletedListener = mReplicationsCompletedListener;
+    }
+
+    public void add(Replicator replicator) {
+        synchronized (replicatorsInProgress) {
+            replicatorsInProgress.add(replicator);
+        }
+    }
+
+    public void remove(Replicator replicator) {
+        synchronized (replicatorsInProgress) {
+            if (replicatorsInProgress.remove(replicator)) {
+                replicator.getEventBus().unregister(this);
```

Comment:
I think we got a bit muddled by the different listeners in play here. I'm confident this is actually ok because:
1. An instance of `this` `ReplicationListener` is registered on each `Replicator`'s own `EventBus` in the loop in `ReplicationPolicyManager.startReplications()`.
1. The `unregister` here in `remove` is the counterpart to that registration for when `stopReplications`  has been called or the replication complete/errored events have fired.

Unlike the replicator's `EventBus` for each `ReplicationPolicyManager` there can only be a single `ReplicationsCompletedListener` set. So I don't think there is any issue with multiple registrations. That said there is currently nothing to stop somebody changing the `ReplicationsCompletedListener` at any given point (or setting it to `null`), but the instance of `ReplicationListener` would not change and it has `null` checks so I don't think this would actually break anything.

However, I do think this would be more clear if the registration/unregistration were done symmetrically. Also, looking at it further I don't think the `inProgress` check is necessary as calling start twice on the same replication is a no-op (unless it already finished, but if that was the case it would have fallen out of the `inProgress` set anyway). The tests fail if I change that because they assert the number of times start/stop were called, but we can still avoid the multiple calls without the `inProgress` method just by returning the set addition/removal result `boolean`.

@tomblench @brynh see 055a8bb69fc43ce931d3759e47f1ebcb11962520